src/js/data/ui-update/index.ts

```diff
@@ -0,0 +1,131 @@
+import { Observable, of, throwError } from "rxjs";
+import {
+  exhaustMap,
+  map,
+  publishReplay,
+  refCount,
+  retry
+} from "rxjs/operators";
+import { RequestResponse } from "@dcos/http-service";
+import { makeExecutableSchema } from "graphql-tools";
+import { DCOSUIUpdateClient, UIVersionResponse } from "dcos-ui-update-client";
+
+import Config from "#SRC/js/config/Config";
+import { UIVersionSchema } from "#SRC/js/data/ui-update/UIVersion";
+
+export interface ResolverArgs {
+  uiUpdateClient: DCOSUIUpdateClient;
+}
+
+export interface GeneralArgs {
+  [key: string]: any;
+}
+
+export interface QueryContext {
+  [key: string]: any;
+}
+
+export interface UIUpdateArgs {
+  newVersion: string;
+}
+
+function isUIUpdateArgs(args: GeneralArgs): args is UIUpdateArgs {
+  return (args as UIUpdateArgs).newVersion !== undefined;
+}
+
+export function resolvers({ uiUpdateClient }: ResolverArgs) {
+  return {
+    UIVersion: {
+      clientBuild() {
+        // @ts-ignore
+        return of(window.DCOS_UI_VERSION);
+      },
+      packageVersion(
+        _parent = {},
+        _args = {},
+        context: { _versionStream: Observable<any> }
+      ) {
+        const version$ = context._versionStream;
+        return version$.pipe(
+          map(
+            ({ response }: RequestResponse<UIVersionResponse>) =>
+              response.packageVersion
+          )
+        );
+      },
+      packageVersionIsDefault(
+        _parent = {},
+        _args = {},
+        context: { _versionStream: Observable<any> }
+      ) {
+        const version$ = context._versionStream;
+        return version$.pipe(
+          map(
+            ({ response }: RequestResponse<UIVersionResponse>) =>
+              response.default
+          )
+        );
+      }
+    },
+    Query: {
+      ui(_parent = {}, _args = {}, context: QueryContext = {}) {
+        context._versionStream = of({}).pipe(
```

Comment:
You could put this stream also just on the top of the function instead of here, so that you don't need to use the context.


```js
export function resolvers({ uiUpdateClient }: ResolverArgs) {
  const version$ = of({})....

  return {
}
```

Reply:
It is, we didn’t want to poll with any of these resolvers so I left it out for now.

Reply:
I think we did not talk about the same thing here.

In this field resolver you create a obervable, which is subscribed to later on by different other field resolvers. While this is a great pattern I think we should try to use the context as little as possible, at best only to pass meta-data for the request from one level to the other. The observable you have here could also be held outside of any resolvers and subscribed to by a normal object reference, which seems to me way less error prone (e.g. eslint will cry if you rename that reference, but not if you mask it through an object). It's no blocker for me, I just wanted to be clear on what I meant before moving on 😇

Reply:
@DanielMSchmidt I don't think we can. Since we are not polling creating it at the top of the function would result in the API only being called a single time.

I put it here so that the stream is created when you make a query that includes `ui`, but is still shared between the `ui` field resolvers. That way we make 1 new request per query, if any of the field resolver's need to call the ui-service API. Other wise the stream is not executed and the API request is not made. At least based on the unit tests and manually testing that is the desired behavior.

Reply:
Ah, damn, now I understand what you mean. I didn't make the connection between your words and what you meant with polling there :D please carry on